### PR TITLE
_TableDatasets allow NaN with optional imputation after scaling

### PR DESCRIPTION
**Pull request by C-nit**
Bugfixes:
 - In case of multiple files with different column order, data is ordered for transformation by delegation to underlying datasets
 - minmax scaling

Features:
 - respect missing features when computing and reducing feature-wise statistics
 - added optional `impute` argument.  
   Imputation of user defined value happens after scaling and does not scew distribution for statistics computation

### CODE REVIEW

File: pytoda/datasets/_csv_statistics.py [hunk 1]
```diff
@@ -27,40 +33,141 @@ def __init__(
...
         self.max = self.min_max_scaler.data_max_
         self.min = self.min_max_scaler.data_min_
         self.mean = self.standardizer.mean_
         self.std = self.standardizer.scale_
 
-    def setup_datasource(self) -> pd.Series:
+    def _reindex(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Missing features are represented with all zeros."""
+        # fill = 0.0
+        # fills with NaN TODO docs
```
Comment (jannisborn) [inline pytoda/datasets/_csv_statistics.py]: I think 0-imputation makes sense but this method looks a bit unfinished to me, especially since the default `fill_value` is `np.nan` and not 0.

File: pytoda/datasets/_csv_statistics.py [hunk 2]
```diff
@@ -102,40 +209,45 @@ def reduce_csv_statistics(
...
+                csv_dataset.notna_count[
+                    csv_dataset.feature_mapping[features].values]
             ) for csv_dataset in csv_datasets
         ]
     )
     # NOTE: reduce max and min
-    maximum = np.array(maximums).max(axis=0)
-    minimum = np.array(minimums).min(axis=0)
+    maximum = np.nanmax(maximums, axis=0)
+    minimum = np.nanmin(minimums, axis=0)
```
Comment (jannisborn) [inline pytoda/datasets/_csv_statistics.py:230]: Are these statistics computed on the imputed or raw values? Imagine a situation where we impute NaNs with 0 (which should be the default), but we only have negative values. Will the maximum be computed correctly here (i.e. 0) or is the imputation applied _after_ the computation of these statistics?

**Review by jannisborn — COMMENTED** [review]
Nice that you spotted this issue. I went through it and found two small things where I wasn't sure that they were right. Overall, Matteo may be the better reviewer for this part of the package.

This PR should close #61, right?

File: pytoda/datasets/_csv_statistics.py (hunk 1 above)
Comment (C-nit) [reply pytoda/datasets/_csv_statistics.py]: yes good catch, this is a TODO I forgot to clean up.

File: pytoda/datasets/_csv_statistics.py (hunk 2 above)
Comment (C-nit) [reply pytoda/datasets/_csv_statistics.py:230]: The imputation is applied _after_ the transformation on purpose: You know more about the data distribution after the transformation and we can reason on imputation values on these scales. I.e. imputing zero after minmax scaling is setting a minimal value, and for standardisation it is imputing a mean value.

In the case you describe I would think that the goal of minmax scaling would be to map them into the range [0, 1]. Imputing zero beforehand will scew this transformation (lets say the values are in [-2, -1], they would end up in [0, 0.5]) and the imputed zeros will be maximal values (1).

If you know more about your initial data distribution and want impute values that impact the statistics, one would have to impute NaNs explicitly beforehand on the files.

The refactor would also allow you to impute values "manually" using the datasets implementation after it's initialisation, though it requires knowledge of the inner API.

Comment (C-nit) [reply pytoda/datasets/_csv_statistics.py:230]: Do you think imputing beforehand instead is a common case that should be supported as well?

File: pytoda/datasets/_csv_statistics.py (hunk 1 above)
Comment (C-nit) [reply pytoda/datasets/_csv_statistics.py]: To be clear, we do impute nothing/np.NaN here on purpose.

File: pytoda/datasets/_csv_statistics.py (hunk 2 above)
Comment (jannisborn) [reply pytoda/datasets/_csv_statistics.py:230]: No no, I think imputing afterwards is the only think that makes sense. I just wanted to double check that the code works in this way :)

File: pytoda/datasets/_csv_statistics.py (hunk 1 above)
Comment (jannisborn) [reply pytoda/datasets/_csv_statistics.py]: Ok, this is surprising since the old behavior was a 0 imputation. What's the rationale of the nans and what happens with the nan values then during the transformation?

Comment (C-nit) [reply pytoda/datasets/_csv_statistics.py]: At this stage the data is read the first time. Here we need to keep any NaN and fill missing features given by the `initial_feature_list` with NaN (making it explicit here, though it is the default).

This way the statistics can be computed correctly ignoring NaN (this required the changes you commented below e.g. from `np.max` to `np.nanmax`. NaNs stay Nans in the transformation. After transformation the actual imputation is performed.

In other words, the old default of imputing 0 here was wrong.

**Comment by C-nit** [comment]
> This PR should close #61, right?  

No it does not address it.
The changes here are motivated by my need to support missing values in a more meaningful manner and that minmax scaling was breaking, not on conditionally only fitting one of the two scalers.

**Review by jannisborn — APPROVED** [review]
